addons/stock/procurement.py

```diff
@@ -1,120 +1,92 @@
+    move_ids = fields.One2many('stock.move', 'procurement_id', string='Moves', help="Moves created by the procurement")
+    move_dest_id = fields.Many2one(comodel_name='stock.move', string='Destination Move', help="Move which caused (created) the procurement")
+    route_ids = fields.Many2many('stock.location.route', 'stock_location_route_procurement', 'procurement_id', 'route_id', string='Preferred Routes', help="Preferred route to be followed by the procurement order. Usually copied from the generating document (SO) but could be set up manually.")
+    warehouse_id = fields.Many2one(comodel_name='stock.warehouse', string='Warehouse', help="Warehouse to consider for the route selection")
+    orderpoint_id = fields.Many2one(comodel_name='stock.warehouse.orderpoint', string='Minimum Stock Rule')
+
+    @api.multi
+    def propagate_cancels(self):
+        self.filtered(lambda proc: proc.rule_id.action == 'move' and proc.move_ids).mapped('move_ids').action_cancel()
         return True
 
-    def cancel(self, cr, uid, ids, context=None):
-        if context is None:
-            context = {}
-        to_cancel_ids = self.get_cancel_ids(cr, uid, ids, context=context)
-        ctx = context.copy()
-        #set the context for the propagation of the procurement cancellation
-        ctx['cancel_procurement'] = True
-        self.propagate_cancels(cr, uid, to_cancel_ids, context=ctx)
-        return super(procurement_order, self).cancel(cr, uid, to_cancel_ids, context=ctx)
-
-    def _find_parent_locations(self, cr, uid, procurement, context=None):
-        location = procurement.location_id
+    @api.multi
+    def cancel(self):
+        to_cancel = self.browse(self.get_cancel_ids()).with_context({'cancel_procurement': True})
+        to_cancel.propagate_cancels()
+        return super(ProcurementOrder, self).cancel()
+
+    @api.multi
+    def _find_parent_locations(self):
+        self.ensure_one()
+        location = self.location_id
         res = [location.id]
         while location.location_id:
             location = location.location_id
-            res.append(location.id)
+            res.append(location.location_id.id)
         return res
 
-    def change_warehouse_id(self, cr, uid, ids, warehouse_id, context=None):
-        if warehouse_id:
-            warehouse = self.pool.get('stock.warehouse').browse(cr, uid, warehouse_id, context=context)
-            return {'value': {'location_id': warehouse.lot_stock_id.id}}
-        return {}
+    @api.multi
+    @api.onchange('warehouse_id')
+    def change_warehouse_id(self):
+        if self.warehouse_id:
+            self.location_id = self.warehouse_id
 
     #Doing assignation, ... in multi
-    def _assign_multi(self, cr, uid, procurements, context=None):
+    @api.model
+    def _assign_multi(self, procurements):
```

Comment:
Here really need decorator ?? because its calling from already migrated code so check it once again. :large_blue_circle: